cloudca/resource_port_forwarding_rule.go

```diff
@@ -0,0 +1,183 @@
+		Create: createPortForwardingRule,
+		Read:   readPortForwardingRule,
+		Delete: deletePortForwardingRule,
+
+		Schema: map[string]*schema.Schema{
+			"service_code": &schema.Schema{
+				Type:        schema.TypeString,
+				Required:    true,
+				ForceNew:    true,
+				Description: "A cloudca service code",
+			},
+			"environment_name": &schema.Schema{
+				Type:        schema.TypeString,
+				Required:    true,
+				ForceNew:    true,
+				Description: "Name of environment where port forwarding rule should be created",
+			},
+			"public_ip_id": &schema.Schema{
+				Type:        schema.TypeString,
+				Required:    true,
+				ForceNew:    true,
+				Description: "The public IP to which these rules should be applied",
+			},
+			"instance_id": &schema.Schema{
+				Type:        schema.TypeString,
+				Optional:    true,
+				ForceNew:    true,
+				Computed:    true,
+				Description: "Name or ID of the instance that this rule should be applied to. If a private IP is not specified, the instance's primary private IP will be selected.",
+			},
+			"private_ip_id": &schema.Schema{
+				Type:        schema.TypeString,
+				Optional:    true,
+				ForceNew:    true,
+				Computed:    true,
+				Description: "The ID of the private IP to bind to. Does not require an instance to be specified.",
+			},
+			"protocol": &schema.Schema{
+				Type:        schema.TypeString,
+				Required:    true,
+				ForceNew:    true,
+				Description: "The protocol that this rule should use (eg. TCP, UDP)",
+			},
+			"private_port_start": &schema.Schema{
+				Type:        schema.TypeInt,
+				Required:    true,
+				ForceNew:    true,
+				Description: "The start of the private port range for this rule",
+			},
+			"private_port_end": &schema.Schema{
+				Type:        schema.TypeInt,
+				Required:    true,
+				ForceNew:    true,
+				Description: "The end of the private port range for this rule",
+			},
+			"public_port_start": &schema.Schema{
+				Type:        schema.TypeInt,
+				Required:    true,
+				ForceNew:    true,
+				Description: "The start of the public port range for this rule",
+			},
+			"public_port_end": &schema.Schema{
+				Type:        schema.TypeInt,
+				Required:    true,
+				ForceNew:    true,
+				Description: "The end of the public port range for this rule",
+			},
+			"public_ip": &schema.Schema{
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+			"private_ip": &schema.Schema{
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+		},
+	}
+}
+
+func validateOneOfPrivateIpIdAndInstanceIdAreProvided(d *schema.ResourceData) error {
+	_, instance_ok := d.GetOk("instance_id")
+	_, private_ip_ok := d.GetOk("private_ip_id")
+
+	if !instance_ok && !private_ip_ok {
+		return fmt.Errorf("At least one of instance_id and private_ip_id must be specified")
+	}
+
+	return nil
+}
+
+func createPortForwardingRule(d *schema.ResourceData, meta interface{}) error {
+	err := validateOneOfPrivateIpIdAndInstanceIdAreProvided(d)
```

Comment:
The server verifies that the private IP belongs to the instance and then behaves as it would if just the private IP was specified.